src/BlockNLPModels.jl

```diff
@@ -0,0 +1,213 @@
+  & \\sum\\limits_{i \\in \\mathcal{B}} A_i x_i = b,
+\\end{aligned}
+where ``\\mathcal{B}`` is the set of variable blocks.
+
+---
+
+    BlockNLPModel()
+
+Initializes an empty `BlockNLPModel`. 
+"""
+mutable struct BlockNLPModel <: AbstractBlockNLPModel
+  problem_size::BlockNLP_Counters
+  blocks::Vector{AbstractBlockModel}
+  linking_constraints::Vector{Union{AbstractLinearLinkConstraint, AbstractNLLinkConstraint}}
+  function BlockNLPModel()
+    return new(
+      BlockNLP_Counters(),
+      Vector{AbstractBlockModel}(),
+      Vector{Union{AbstractLinearLinkConstraint, AbstractNLLinkConstraint}}()
+    )
+  end
+end
+
+"""
+  add_block(
+    block_nlp::AbstractBlockNLPModel, 
+    nlp::AbstractNLPModel
+  )
+
+Adds a block subproblem to a `BlockNLPModel`.
+The `BlockNLPModel` needs to be initialized before this function can be called.
+
+# Arguments
+
+- `block_nlp::AbstractBlockNLPModel`: name of the BlockNLPModel to which a subproblem is to be added.
+- `nlp::AbstractNLPModel`: the subproblem.
+"""
+function add_block(block_nlp::AbstractBlockNLPModel, nlp::AbstractNLPModel)
+  block_nlp.problem_size.block_counter += 1
+
+  temp = deepcopy(block_nlp.problem_size.var_counter)+1
+  block_nlp.problem_size.var_counter += nlp.meta.nvar
+  var_idx = temp:block_nlp.problem_size.var_counter
+
+  temp = deepcopy(block_nlp.problem_size.con_counter)+1 
+  block_nlp.problem_size.con_counter += nlp.meta.ncon
+  con_idx = temp:block_nlp.problem_size.con_counter
+  push!(block_nlp.blocks, AbstractBlockModel(nlp.meta, Counters(), 
+  nlp, block_nlp.problem_size.block_counter, var_idx, con_idx, Vector{Int}()))
+end
+
+"""
+    add_links(block_nlp::AbstractBlockNLPModel, 
+    n_constraints::Int,   
+    links::Union{Dict{Int, Array{Float64, 2}}, Dict{Int, Vector{Float64}}}, 
+    constants::Union{AbstractVector, Float64})
+
+Creates linear link constraint(s) between different subproblems of the BlockNLPModel.
+This function must be called after subproblems have already been added.
+
+# Arguments
+
+- `block_nlp::AbstractBlockNLPModel`: The BlockNLPModel to which the constraint(s) is to be added.  
+- `n_constraints::Int`: Number of link constraints. 
+- `links::Union{Dict{Int, Array{Float64, 2}}, Dict{Int, Vector{Float64}}}`: coefficients of the block matrices that link different blocks,
+- `constants::Union{AbstractVector, Float64}`: RHS vector
+"""
+# 
+function add_links(block_nlp::AbstractBlockNLPModel, n_constraints::Int, 
+  links::Union{Dict{Int, Array{Float64, 2}}, Dict{Int, Vector{Float64}}}, 
+  constants::Union{AbstractVector, Float64})
+  
+  linking_blocks = Vector{SparseMatrixCSC{Float64, Int}}(undef, block_nlp.problem_size.block_counter)
+  RHS_vector = Vector{Float64}(undef, n_constraints)
+
+  # Initialize linking_blocks with empty matrices
+  for i in 1:block_nlp.problem_size.block_counter
+    linking_blocks[i] = spzeros(n_constraints, block_nlp.blocks[i].meta.nvar)
+  end
```

Comment:
In that case, we can instantiate `linking_blocks` directly as:

```julia
linking_blocks = [spzeros(n_constraints, block_nlp.blocks[i].meta.nvar) for i in 1:block_nlp.problem_size.block_counter]
```
(in general it is better to avoid dealing with `undef` array)